src/core/lv2/Lv2Proc.cpp

```diff
@@ -147,31 +182,89 @@ void Lv2Proc::copyModelsFromCore()
 
 	struct Copy : public Lv2Ports::Visitor
 	{
+		ringbuffer_reader_t<MidiInputEvent>* m_midiInputReader;
+		
 		void visit(Lv2Ports::Control& ctrl) override
 		{
-			if (ctrl.m_flow == Lv2Ports::Flow::Input)
-			{
-				FloatFromModelVisitor ffm;
-				ffm.m_scalePointMap = &ctrl.m_scalePointMap;
-				ctrl.m_connectedModel->accept(ffm);
-				ctrl.m_val = ffm.m_res;
-			}
+			FloatFromModelVisitor ffm;
+			ffm.m_scalePointMap = &ctrl.m_scalePointMap;
+			ctrl.m_connectedModel->accept(ffm);
+			ctrl.m_val = ffm.m_res;
 		}
 		void visit(Lv2Ports::Cv& cv) override
 		{
-			if (cv.m_flow == Lv2Ports::Flow::Input)
+			FloatFromModelVisitor ffm;
+			ffm.m_scalePointMap = &cv.m_scalePointMap;
+			cv.m_connectedModel->accept(ffm);
+			// dirty fix, needs better interpolation
+			std::fill(cv.m_buffer.begin(), cv.m_buffer.end(), ffm.m_res);
+		}
+		void visit(Lv2Ports::AtomSeq& atomPort) override
+		{
+			lv2_evbuf_reset(atomPort.m_buf.get(), true);
+		}
+	} copy;
+
+	// initialize copier helper variables
+	copy.m_midiInputReader = &m_midiInputReader;
+
+	// feed each input port with the respective data from the LMMS core
+	for (const std::unique_ptr<Lv2Ports::PortBase>& port : m_ports)
+	{
+		if (port->m_flow == Lv2Ports::Flow::Input)
+		{
+			port->accept(copy);
+		}
+	}
+
+	// send pending MIDI events to atom port
+	if(m_midiIn)
+	{
+		LV2_Evbuf_Iterator iter = lv2_evbuf_begin(m_midiIn->m_buf.get());
+		if(m_midiIn->flags & Lv2Ports::AtomSeq::FlagType::Midi)
```

Comment:
This condition is checked when initialising `m_midiIn`, so doesn't need repeating here.